Support local path "replace" in go.mod

fabric-peter opened the pull request:
Issue #14996 continues to reproduce in 2.24.1: In Go, when multiple modules reference each other in a monorepo, they do so with "require" and "replace" lines in `go.mod`, e.g. if `module_a/package_a` depends on `module_b/package_b`, the `module_a/go.mod` file will contain two lines:
```
require module_b v0.0.0
replace module_b => ../module_b
```

This, however, causes pants to fail as described in issue #19463 and in issue #14996.

It is possible to work around this issue to get the pants build to pass: remove the require and replace lines and use the `dependencies` field in the `go_package` target to manage the dependency, but this unfortunately breaks the Go tools. `go mod ...`, `go test ...` etc. no longer run since they cannot resolve module_b without it being specified in `go.mod`.

This patch adds proper interpretation of a `replace` clause that lacks a Version: interpret the dependency as a first-party module within the same repository and not a third-party module that requires remote fetching. This version of the `replace` clause is described at https://go.dev/ref/mod#go-mod-file-replace
> If the path on the right side of the arrow is an absolute or relative path
> (beginning with ./ or ../), it is interpreted as the local file path to the
> replacement module root directory, which must contain a go.mod file. The
> replacement version must be omitted in this case.

This patch does not add dependency inference across these references, but merely allows the explicit dependencies to function in both pants and go.

tgolsson commented:
Hey @fabric-peter,

Thanks for the PR! While I can believe this fixes the issue, it'd be fantastic if you could add a test-case for this behavior in `third_party_pkg_test.py`. Preferably one that provably fails on main, so we can prove the issue is resolved. I'll unblock CI in the meantime and we can see if anyting *breaks* from this, though I'd assume you've tested this on something locally.

tgolsson commented:
(Ah, and as the CI helpfully points out: We'll need you to add a note in the release notes (`2.26.x.md`).)

fabric-peter commented:
Thank you @tgolsson ! I've added a recursion test, verified that it fails in main but passes in this branch. I've also added a release not following the format from a different fix in the same release notes. Please let me know if I need to reword it or anything.

Also, is there any chance at merging this back into the 2.24.x branch at some point, or once this is merged do I need to start running my builds off of the current main to get the fix?

tgolsson commented:
A backport to 2.24 can probably be done. Can probably get it into a 2.25.1 as well; 2.25.0 is slated for next week as I recall so probably don't want to rambo-merge for that as we've gone through the rc process already.

fabric-peter commented:
Getting it into 2.25 would be amazing, but I totally understand holding it for 2.25.1! If that complicates things for getting this into 2.24.x I can exercise patience :) (aka work off my fork for the near term)

benjyw commented:
Thanks for this @fabric-peter! It's been an open wound for a while now, but people band-aided over it. I'll take a look and also give a holler to @tdyas , who knows this code best.

benjyw commented on src/python/pants/backend/go/util_rules/third_party_pkg.py:
```diff
@@ -229,6 +229,10 @@ async def analyze_module_dependencies(request: ModuleDescriptorsRequest) -> Modu
         if "Main" in mod_json and mod_json["Main"]:
             continue
 
+        # Skip first-party modules referenced from other first-party modules.
```
Should we add a TODO here about using the replacement value for dep inference? Is that a thing we will want to do, or is this dep inferrable some other way already?

fabric-peter replied:
It is not yet able to infer through the replacement value, but that would be a great addition in the future! I'll add a TODO. For now, I've been using explicit references between packages that cross the module boundaries, but automatic would be much cleaner.

```suggestion
        # Skip first-party modules referenced from other first-party modules.
        # TODO These cross-module references could be used for dependency inference
```

benjyw reviewed (approved):
Looks great! Thanks for this. Just one question about how this should affect dep inference. And I'd also like @tdyas to have a quick look at this.

tdyas commented on src/python/pants/backend/go/util_rules/third_party_pkg.py:
```diff
@@ -229,6 +229,11 @@ async def analyze_module_dependencies(request: ModuleDescriptorsRequest) -> Modu
             continue
 
+        # Skip first-party modules referenced from other first-party modules.
+        # TODO These cross-module references could be used for dependency inference
```
Maybe open a Pants issue and put the issue number here next to the TODO so we can track the suggested dependency inference work?

fabric-peter replied:
Thanks! Opened Issue #22097 and referenced it in the TODO here.

tdyas reviewed (commented):
Is there an example of how to actually use the local `replace` support that this PR fixes?

tdyas commented:
> A backport to 2.24 can probably be done. Can probably get it into a 2.25.1 as well; 2.25.0 is slated for next week as I recall so probably don't want to rambo-merge for that as we've gone through the rc process already.
> 
> (@tdyas fyi ^ unless you think it's fine to ninja into 2.25 just letting you know to not hold it for this)

I'd prefer we wait until after 2.25.0 is released this week before back porting (for release in 2.25.1). This bug fix is not fixing a regression so it should not block 2.25.0 final release.

fabric-peter commented:
> Is there an example of how to actually use the local `replace` support that this PR fixes?

Unfortunately I do not have an open source example, but the official Go documentation on modules describes the usage: https://go.dev/doc/modules/managing-dependencies#local_directory

benjyw commented:
@tdyas do you concur that this is good to merge?

tdyas commented:
> @tdyas do you concur that this is good to merge?

Yes. I just released 2.25.0, so the chery pick back to the `2.25.x` will be released in 2.25.1 as desired.

benjyw commented:
Thanks @fabric-peter !

fabric-peter commented:
> Thanks @fabric-peter !

Thank you so much @benjyw for helping me get this in!
